chapter-4.qmd

```diff
@@ -0,0 +1,246 @@
+---
+title: "Data transformation"
+---
+
+## Prerequisites
+
+```{r}
+library(nycflights13)
+library(tidyverse)
+#> ── Attaching packages ──────────────────────────────────── tidyverse 1.3.2 ──
+#> ✔ ggplot2 3.4.0           ✔ purrr   1.0.1.9000 
+#> ✔ tibble  3.1.8           ✔ dplyr   1.0.99.9000
+#> ✔ tidyr   1.2.1.9001      ✔ stringr 1.5.0      
+#> ✔ readr   2.1.3           ✔ forcats 0.5.2      
+#> ── Conflicts ─────────────────────────────────────── tidyverse_conflicts() ──
+#> ✖ dplyr::filter() masks stats::filter()
+#> ✖ dplyr::lag()    masks stats::lag()
+```
+
+##Solutions
+
+## 4.2.5 Exercises
+
+1.  
+
+```{=html}
+<!-- -->
+```
+a.  
+
+```{r}
+filter(flights,arr_delay >= 2) |> 
+  arrange(desc(arr_delay))
+```
+
+b.  Flew to houston (IAH or HOU)
+
+```{r}
+flights |>
+  filter(dest %in% c("IAH","HOU"))
+```
+
+c.  
+
+```{r}
+flights |>
+  filter(carrier %in% c("UA","DL","AA"))
+```
+
+d.  
+
+```{r}
+flights |>
+  filter(month %in% c(7,8,9))
+```
+
+e.  
+
+```{r}
+flights |> 
+  filter(dep_delay >= 0,arr_delay < -2)
+```
+
+f.  
+
+```{r}
+flights |> 
+  filter(dep_delay <= -1,arr_delay > 30)
+```
+
+2.  
+
+```{r}
+flights |> 
+  arrange(dep_delay) |> 
+  arrange(time_hour)
+```
+
+3.  
+
+```{r}
+flights |> 
+  arrange(desc(arr_delay))
+```
+
+4.Yes
+
+```{r}
+flights |> 
+  arrange(desc(day)) |> 
+  arrange(desc(month))
+```
+
+5.  
+
+```{r}
+flights |> 
+  arrange(distance) 
+```
+
+```{r}
+flights |> 
+  arrange(desc(distance))
+```
+
+6.  I don't think it matters the two functions would go over each other with no problems either way.
+
+## 4.3.5 Exercises
+
+1.  No they don't match.
+
+    ```{r}
+    flights |> 
+      mutate(timearrdep = arr_time - dep_time)
+    ```
+
+2.  They show when a plane left when it was supposed to leave and if there was a delay.
+
+3.  
+
+```{r}
+flights |> 
+  select(dep_time, dep_delay, arr_time, arr_delay)
+```
+
+```{r}
+select(flights, dep_time:arr_delay, -sched_dep_time,-sched_arr_time)
+```
+
+4.  Nothing happens, you just get the variable once.
+
+```{r}
+flights |> 
+  select(dep_time, dep_time)
+```
+
+5.  You ask if `any_of()` these variables have a certain thing you are looking for.
+
+```{r}
+variables <- c("year", "month", "day", "dep_delay", "arr_delay")
+```
+
+6.  No it doesn't surprise me.it looks like case doesn't matter.i can change time to dep.
+
+```{r}
+select(flights, contains("TIME"))
+```
+
+## 4.4.6 Exercises
+
+1.  It looks like 9E has the worst delays.
+
+```{r}
+flights |> 
+  group_by(carrier, dep_delay) |> 
+  summarize(n())
+
+flights |> 
+  arrange(desc(dep_delay))
+```
+
+2.  
```

Comment:
We're looking for most delayed by destination, so group by destination first and then arrange by dep_delay in decreasing order.